csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Assets.cs

```diff
@@ -0,0 +1,162 @@
+        }
+
+        /// <summary>
+        /// Add the package dependencies from the assets file to dependencies.
+        /// 
+        /// Parse a part of the JSon assets file and add the paths
+        /// to the dependencies required for compilation (and collect
+        /// information about used packages).
+        ///
+        /// Example:
+        /// {
+        ///   "Castle.Core/4.4.1": {
+        ///     "type": "package",
+        ///     "compile": {
+        ///       "lib/netstandard1.5/Castle.Core.dll": {
+        ///         "related": ".xml"
+        ///        }
+        ///     }
+        ///   },
+        ///   "Json.Net/1.0.33": {
+        ///     "type": "package",
+        ///     "compile": {
+        ///       "lib/netstandard2.0/Json.Net.dll": {}
+        ///     },
+        ///     "runtime": {
+        ///       "lib/netstandard2.0/Json.Net.dll": {}
+        ///     }
+        ///   }
+        /// }
+        /// 
+        /// Returns dependencies
+        ///   Required = {
+        ///     "castle.core/4.4.1/lib/netstandard1.5/Castle.Core.dll",
+        ///     "json.net/1.0.33/lib/netstandard2.0/Json.Net.dll"
+        ///   }
+        ///   UsedPackages = {
+        ///     "castle.core",
+        ///     "json.net"
+        ///   }
+        /// </summary>
+        private Dependencies AddPackageDependencies(JObject json, Dependencies dependencies)
+        {
+            // If there are more than one framework we need to pick just one.
+            // To ensure stability we pick one based on the lexicographic order of
+            // the framework names.
+            var references = json
+                .GetProperty("targets")?
+                .Properties()?
+                .MaxBy(p => p.Name)?
+                .Value
+                .ToObject<Dictionary<string, ReferenceInfo>>();
+
+            if (references is null)
+            {
+                progressMonitor.LogDebug("No references found in the targets section in the assets file.");
+                return dependencies;
+            }
+
+            // Find all the compile dependencies for each reference and
+            // create the relative path to the dependency.
+            return references
+                .Aggregate(dependencies, (deps, r) =>
+                {
+                    var info = r.Value;
+                    var name = r.Key.ToLowerInvariant();
+                    if (info.Type != "package")
+                    {
+                        return deps;
+                    }
+
+                    // If this is a .NET framework reference then include everything.
+                    return netFrameworks.Any(framework => name.StartsWith(framework))
+                        ? deps.Add(name, "")
+                        : info
+                            .Compile
+                            .Aggregate(deps, (d, p) => d.Add(name, p.Key));
+                });
```

Comment:
I find this hard to read; `dependencies`, `deps`, and `d` all refer to the same variable, don't they?
Is this code equivalent to the below? (I find the below a lot easier to read)

```C#
foreach (var r in references)
{
    var info = r.Value;
    var name = r.Key.ToLowerInvariant();
    if (info.Type != "package")
    {
        continue;
    }
    // If this is a .NET framework reference then include everything.
    if (netFrameworks.Any(name.StartsWith))
    {
        dependencies.Add(name, "");
        continue;
    }
    // Otherwise only include the files mentioned in the compile section.
    foreach (var p in info.Compile)
    {
        dependencies.Add(name, p.Key);
    }
}
return dependencies;
```

Reply:
Yes, I agree that `Aggregate` is not the most suitable way of doing this (I usually think in terms of functional programming and aggregation is just "folding", but since the data is collected as a side effect it doesn't make sense to use aggregate).
I know, we have a bit different preference when comes to implementing loops and I will try and do this using the `ForeEach` extension method instead - then the implementation also become similar to your suggestion :-)
Or should I just cave in and start using the `foreach` more? :-D

Reply:
You should probably try to convince me that I should start using `ForEach` more. :-)

I think it doesn't matter much which variant is used, I prefer readability over shortness. Lambdas oftentimes help with readability, but I'm not (yet) convinced that this is the case with simple `foreach` loops. Also, with `ForEach` I could always argue that there's an unnecessary lambda allocation.

Reply:
Haha, I will try 😸 

I like to avoid using `continue` and `break` statements in `foreach` loops - it doesn't take that many of those before the readability starts to deteriorate and they have a tendency to start popping up as the code becomes more complicated. In that aspect I also prefer the `return` statement instead of a `continue` statement, which is what we will get, if we use lambdas (this also makes it easier to re-factor the entire lambda out into a separate method, which encourages refactors when needed).
Furthermore, it is not possible to use `break` inside `ForEach` (ForEach means *for each* (unless an exception is thrown 😄 )). It is my impression that `break` usually means that we are searching for a specific element or collecting elements up until a specific condition or something similar and at the same time we and may or may not *map* these elements as well (in one go): I generally prefer to do filtering and then processing (if possible). The approach of doing lots of things at the same time also has a tendency to make the code more complicated in the long run.

In terms of performance, I agree that it is sometimes better to use the native C# `foreach` construct for performance reasons and I am sure the performance difference between `foreach` and `ForEach` is measurable if it is a loop that is executed 100k+ times. However, in most cases the difference in performance is completely insignificant.